io/io/inc/ROOT/TBufferMerger.hxx

```diff
@@ -0,0 +1,132 @@
+
+#ifndef ROOT_TBufferMerger
+#define ROOT_TBufferMerger
+
+#include "TMemFile.h"
+
+#include <condition_variable>
+#include <memory>
+#include <mutex>
+#include <queue>
+#include <thread>
+
+class TArrayC;
+class TBufferFile;
+
+namespace ROOT {
+namespace Experimental {
+
+class TBufferMergerFile;
+
+/**
+ * \class TBufferMerger TBufferMerger.hxx
+ * \ingroup IO
+ *
+ * TBufferMerger is a class to facilitate writing data in
+ * parallel from multiple threads, while writing to a single
+ * output file. Its purpose is similar to TParallelMergingFile,
+ * but instead of using processes that connect to a network
+ * socket, TBufferMerger uses threads that each write to a
+ * TBufferMergerFile, which in turn push data into a queue
+ * managed by the TBufferMerger.
+ */
+
+class TBufferMerger {
+public:
+   /** TBufferFile has no default constructor */
+   TBufferMerger() = delete;
+
+   /** TBufferFile has no copy constructor */
+   TBufferMerger(const TBufferMerger &) = delete;
+
+   /** Constructor
+    * @param name Output file name
+    * @param option Output file creation options
+    * @param ftitle Output file title
+    * @param compression Output file compression level
+    */
+   TBufferMerger(const char *name, Option_t *option = "RECREATE", const char *ftitle = "", Int_t compress = 1);
+
+   /** Destructor */
+   ~TBufferMerger();
+
+   /** Returns a TBufferMergerFile to which data can be written.
+    *  At the end, all TBufferMergerFiles get merged into the output file.
+    */
+   std::shared_ptr<TBufferMergerFile> GetFile();
+
+   friend class TBufferMergerFile;
+
+private:
+   void Push(TBufferFile *buffer);
+   void Listen();
+
+   std::mutex fFilesMutex;                                       //< Mutex used to lock fAttachedFiles
+   std::mutex fQueueMutex;                                       //< Mutex used to lock fQueue
+   std::mutex fWriteMutex;                                       //< Mutex used for the condition variable
+   std::condition_variable fCV;                                  //< Condition variable used to wait for data
+   std::queue<TBufferFile *> fQueue;                             //< Queue to which data is pushed and merged
+   std::unique_ptr<TFile> fFile;                                 //< Output file, owned by the user
+   std::unique_ptr<std::thread> fMergingThread;                  //< Worker thread that writes to disk
+   std::vector<std::weak_ptr<TBufferMergerFile>> fAttachedFiles; //< Attached files
+};
+
+/**
+ * \class TBufferMerger TBufferMerger.hxx
+ * \ingroup IO
+ *
+ * A TBufferMergerFile is similar to a TMemFile, but when data
+ * is written to it, it is appended to the TBufferMerger queue.
+ * The TBufferMerger merges all data into the output file on disk.
+ */
+
+class TBufferMergerFile : public TMemFile {
+private:
+   TBufferMerger &fMerger;              //< TBufferMerger this file is attached to
+   std::unique_ptr<TArrayC> fClassSent; //< StreamerInfo this file has already written
```

Comment:
This does not need to be addressed before merging.

The main difference between std::bitset and TBits is that bitset has a fixed size at compile time while TBits is expandable (and some operation in TBits are (or at least used to be) faster).  Here you can not predict the maximum number of TStreamerInfo involved in storing data in this file.

The reason it is a TArrayC is that at the time the original version of this code was written (circa 2000), TBits did not exist yet ... i.e. this code is copy/paste per se from TFile::WriteStreamerInfo.

The purpose is solely to keep track of whether a StreamerInfo, refer to by its index has been sent over ...

... right but that does not make any sense when the client and the  server are in the same process ....

So actually WriteStreamerInfo and any use of fClassSent can go away ....